packages/frontend/src/components/rich_text_editor/rich_text_editor.tsx

```diff
@@ -0,0 +1,361 @@
+                    ) {
+                        return true;
+                    }
+
+                    setEditorFocused(false);
+                    return false;
+                },
+            },
+            clipboardTextSerializer: (slice) => {
+                return mathSerializer.serializeSlice(slice);
+            },
+        });
+
+        if (props.ref) {
+            props.ref(view);
+        }
+
+        setHeadingLevel(activeHeading(view.state, schema));
+        setMarkStates(activeMarks(view.state, schema));
+
+        setMenuControls({
+            onBoldClicked: () => toggleMark(schema.marks.strong)(view.state, view.dispatch),
+            onItalicClicked: () => toggleMark(schema.marks.em)(view.state, view.dispatch),
+            onLinkClicked: null,
+            onBlockQuoteClicked: () => turnSelectionIntoBlockquote(view.state, view.dispatch, view),
+            onToggleOrderedList: null,
+            onToggleNumberedList: null,
+            onIncreaseIndent: null,
+            onDecreaseIndent: null,
+            onHeadingClicked: (level: number) => {
+                if (level === 0) {
+                    // paragraph
+                    setBlockType(schema.nodes.paragraph)(view.state, view.dispatch);
+                } else {
+                    // heading
+                    setBlockType(schema.nodes.heading, { level })(view.state, view.dispatch);
+                }
+            },
+            onMathClicked: () => {
+                const cmd = insertMathDisplayCmd(schema.nodes.math_display);
+                cmd(view.state, view.dispatch);
+            },
+        });
+
+        onCleanup(() => view.destroy());
+    });
+
+    return (
+        <div class="prosemirror">
+            <Show when={isEditorFocused() || isMenuActive()}>
+                <div
+                    ref={menuRoot}
+                    onFocusIn={() => setMenuActive(true)}
+                    onFocusOut={() => setMenuActive(false)}
+                >
+                    <MenuBar {...menuControls()} {...markStates()} headingLevel={headingLevel()} />
+                </div>
+            </Show>
+            <div class="rich-text-editor" ref={editorRoot} />
+        </div>
+    );
+};
+
+function activeMarks(state: EditorState, schema: CustomSchema): MarkStates {
+    const isBoldActive = isMarkActive(state, schema.marks.strong);
+    const isEmActive = isMarkActive(state, schema.marks.em);
+
+    return { isBoldActive, isEmActive };
+}
+
+function richTextEditorKeymap(schema: CustomSchema, props: RichTextEditorOptions) {
+    const bindings: { [key: string]: Command } = {};
+
+    bindings["Mod-b"] = toggleMark(schema.marks.strong);
+    bindings["Mod-i"] = toggleMark(schema.marks.em);
+    bindings["Mod-m"] = insertMathDisplayCmd(schema.nodes.math_display);
+
+    if (props.deleteBackward) {
+        bindings["Backspace"] = doIfEmpty(props.deleteBackward);
```

Comment:
Once we have inline math we probably want to add `mathBackspaceCmd` here: https://github.com/benrbray/prosemirror-math?tab=readme-ov-file#plugins